server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java

```diff
@@ -104,42 +103,54 @@ public void close() {
             final TimeUnit timeUnit,
             final CheckedRunnable<E> onBlocked) throws InterruptedException, TimeoutException, E {
         delayOperations();
-        try {
-            doBlockOperations(timeout, timeUnit, onBlocked);
+        try (Releasable ignored = acquireAll(timeout, timeUnit)) {
+            onBlocked.run();
         } finally {
             releaseDelayedOperations();
         }
     }
 
     /**
-     * Immediately delays operations and on another thread waits for in-flight operations to finish and then executes {@code onBlocked}
-     * under the guarantee that no new operations are started. Delayed operations are run after {@code onBlocked} has executed. After
-     * operations are delayed and the blocking is forked to another thread, returns to the caller. If a failure occurs while blocking
-     * operations or executing {@code onBlocked} then the {@code onFailure} handler will be invoked.
+     * Immediately delays operations and on another thread waits for in-flight operations to finish and then acquires all permits. When all
+     * permits are acquired, the provided {@link ActionListener} is called under the guarantee that no new operations are started. Delayed
+     * operations are run once the {@link Releasable} is released or if a failure occurs while acquiring all permits; in this case the
+     * {@code onFailure} handler will be invoked after delayed operations are released.
      *
-     * @param timeout   the maximum time to wait for the in-flight operations block
-     * @param timeUnit  the time unit of the {@code timeout} argument
-     * @param onBlocked the action to run once the block has been acquired
-     * @param onFailure the action to run if a failure occurs while blocking operations
-     * @param <E>       the type of checked exception thrown by {@code onBlocked} (not thrown on the calling thread)
+     * @param onAcquired {@link ActionListener} that is invoked once acquisition is successful or failed
+     * @param timeout    the maximum time to wait for the in-flight operations block
+     * @param timeUnit   the time unit of the {@code timeout} argument
      */
-    <E extends Exception> void asyncBlockOperations(
-            final long timeout, final TimeUnit timeUnit, final CheckedRunnable<E> onBlocked, final Consumer<Exception> onFailure) {
+    public void asyncBlockOperations(final ActionListener<Releasable> onAcquired, final long timeout, final TimeUnit timeUnit)  {
         delayOperations();
         threadPool.executor(ThreadPool.Names.GENERIC).execute(new AbstractRunnable() {
+
+            final AtomicBoolean released = new AtomicBoolean(false);
+
             @Override
             public void onFailure(final Exception e) {
-                onFailure.accept(e);
+                try {
```

Comment:
we have this pattern all over the place in ES and here twice in one file where we want to run some code once and only once. Maybe it would be nice to have a simple `RunOnce` class that takes a closure?

Reply:
Good suggestion, I'll do this in a follow up PR

Reply:
++